File: app/src/main/java/com/originb/inkwisenote2/functionalUtils/Try.java
package com.originb.inkwisenote2.functionalUtils;

import com.originb.inkwisenote2.common.Logger;

import java.util.Optional;
import java.util.concurrent.Callable;

public class Try<T> {

    private Logger logger;
    private Callable<T> callable;
    private Exception exception;

    private String exceptionMessage = "Failed to execute callable";

    private Try(Callable<T> callable, Logger logger) {
        this.callable = callable;
        this.logger = logger;
    }


    public static <T> Try<T> to(Callable<T> callable, Logger logger) {
        return new Try<T>(callable, logger);
    }

    public static <T> Try<T> to(Runnable runnable, Logger logger) {
        Callable<T> callable = () -> {
            runnable.run();
            return null;
        };
        return new Try<T>(callable, logger);
    }

    public Try<T> logIfError(String exceptionMessage) {
        this.exceptionMessage = exceptionMessage;
        return this;
    }

    public Optional<T> get() {
        try {
            return Optional.ofNullable(callable.call());
        } catch (Exception e) {
            exception = e;
            logger.exception( exceptionMessage, exception);
        }

        return Optional.empty();
    }
}

File: app/src/main/java/com/originb/inkwisenote2/functionalUtils/Try.kt
package com.originb.inkwisenote2.functionalUtils

import com.originb.inkwisenote2.common.Logger
import java.util.*
import java.util.concurrent.Callable

class Try<T> private constructor(private val callable: Callable<T?>, private val logger: Logger) {
    private var exception: Exception? = null

    private var exceptionMessage: String? = "Failed to execute callable"


    fun logIfError(exceptionMessage: String?): Try<T?> {
        this.exceptionMessage = exceptionMessage
        return this
    }

    fun get(): Optional<T?> {
        try {
            return Optional.ofNullable<T?>(callable.call())
        } catch (e: Exception) {
            exception = e
            logger.exception(exceptionMessage, exception)
        }

        return Optional.empty<T?>()
    }

    companion object {
        fun <T> to(callable: Callable<T?>, logger: Logger): Try<T?> {
            return Try<T?>(callable, logger)
        }

        @JvmStatic
        fun <T> to(runnable: Runnable, logger: Logger): Try<T?> {
            val callable: Callable<T?> = Callable {
                runnable.run()
                null
            }
            return Try<T?>(callable, logger)
        }
    }
}

